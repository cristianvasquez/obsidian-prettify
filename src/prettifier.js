const gfm = require('remark-gfm')
const unified = require('unified')
const parse = require('remark-parse')
const images = require('remark-images')
const frontmatter = require('remark-frontmatter');
const table_writer = require('./table-writer');
const stringify = require('./stringify')
const remark = unified().use(parse).freeze()
const {NEW_HEADER_TEMPLATE} = require('./constants');
const math = require('remark-math')
const katex = require('rehype-katex')
const moment = require('moment')

function prettifier(
    content,
    {
        bullet = '-',
        emphasis = '_',
        rule = '-',
        listItemIndent = 'one', // one, mixed, tab

        createHeaderIfNotPresent: createHeaderIfNotPresent = false,
        newHeaderTemplate = NEW_HEADER_TEMPLATE,
        updateHeader = true,
        currentMoment = moment(),
    } = {}
) {
    let result = remark()
        .use(gfm)

    result = result.use(frontmatter)

    if (createHeaderIfNotPresent || updateHeader) {
        result = result.use(table_writer, {
            createHeaderIfNotPresent: createHeaderIfNotPresent,
            newHeaderTemplate: newHeaderTemplate,
            updateHeader: updateHeader,
            currentMoment: currentMoment
        })
    }

    result
        .use(math)
        .use(katex)
        .use(images)
        .use(stringify, {
            bullet: bullet,
            emphasis: emphasis,
            rule: rule,
            listItemIndent: listItemIndent
        })

    return result.process(content)
}

module.exports = prettifier
