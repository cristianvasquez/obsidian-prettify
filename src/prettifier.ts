import stringify from "./stringify";
import frontmatter from "remark-frontmatter";
import frontmatter_writer from './frontmatter-writer'
import type { MarkdownPrettifierOptions, FrontMatterOptions, FontmatterInput } from "./domain";

import {wikiLinkPlugin} from "./wikilinks";
import gfm from 'remark-gfm';
import math from 'remark-math'
import remark from "remark";


import { NEW_HEADER_TEMPLATE } from "./constants";

import moment from 'moment'

function prettifier(
  content: string,
  options: MarkdownPrettifierOptions = {    
    bullet: "-",
    fence: "~",
    emphasis: "_",
    rule: "-",
    fences: true,
    listItemIndent: "one",
    tightDefinitions:false,    
    incrementListMarker: true,
    createHeaderIfNotPresent: false,
    newHeaderTemplate: NEW_HEADER_TEMPLATE,
    updateHeader: true,
  },
  frontMatterData: FontmatterInput = { today: moment(), tags: [] }
) {

  let processor = remark()

  if (options.createHeaderIfNotPresent || options.updateHeader) {
    processor = processor
      .use(frontmatter)
      .use(frontmatter_writer, options as FrontMatterOptions, frontMatterData);
  }

  const stringifyOptions = Object.assign({}, options)

  if (!stringifyOptions.newlinesAroundHeadings) {
    stringifyOptions.join = [
      //@ts-ignore
      (left, right) => {
        if (left.type === 'heading' || right.type === 'heading') {
          return 0;
        }
        return 1;
      },
    ];
  }

  processor = processor
    .use(wikiLinkPlugin)
    .use(math)
    .use(gfm)
    // @ts-ignore
    .use(stringify, stringifyOptions);
    
  return processor.process(content);
}

export default prettifier
