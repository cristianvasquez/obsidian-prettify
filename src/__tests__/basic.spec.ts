import prettifier from '../prettifier'

test("Basic functionality: a formatted table", () => {
    const content = `
**A**|**B**|**C**
|---:|:---|---|
|a |b |c|
|x |y |z|`;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Basic functionality: lists use * bullets", () => {
    const content = `
- a
- b
  - c
  - d
    - e
    - f
            `;
    return prettifier(content, {bullet: "*"}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Basic functionality: lists use - bullets", () => {
    const content = `
* a
* b
  * c
  * d
    * e
    * f
            `;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Basic functionality: alternating keeps information of different codification", () => {
    const content = `
- a
- b
  * c
  - d
    - e
    * f
            `;
    return prettifier(content, {bullet: "*"}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Basic functionality: enumerate lists", () => {
    const content = `
1. foo
  a. aaa
  b. bbb
  c. ccc
1. bar
1. baz
            `;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Basic functionality: enumerate lists", () => {
    const content = `
1. foo
    1. aaa
    1. bbb
    4. ccc
1. bar
1. baz
            `;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Basic functionality: dont break obsidian links", () => {
    const content = `[[respect this]]`;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Embed youtube", () => {
    const content = `
https://www.youtube.com/watch?v=B6rKUf9DWRI
            `;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Embed image", () => {
    const content = `
https://cyberculturesblog.files.wordpress.com/2019/01/Memex.jpg
            `;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Indent size", () => {
    const content = `
- foo
-  foo
-   foo
-  foo
            `;
    return prettifier(content, {listItemIndent: "one"}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Indent size", () => {
    const content = `
- foo
-  foo
-   foo
-  foo
            `;
    return prettifier(content, {listItemIndent: "tab"}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test('Newlines around headings: enabled', () => {
    const content = `
# Heading
Content
## Subheading
  `;
    return prettifier(content, {newlinesAroundHeadings: true}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Basic functionality: no spread", () => {
    const content = `
* one
* two
            `;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test("Quoted test", () => {
    const content = `
> quoted test
`;
    return prettifier(content, {}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });
});

test('Newlines around headings: disabled', () => {
    const content = `
# Heading
Content
## Subheading
  `;
    return prettifier(content, {newlinesAroundHeadings: false}, {}).then((data) => {
        expect(data.contents).toMatchSnapshot();
    });


});
