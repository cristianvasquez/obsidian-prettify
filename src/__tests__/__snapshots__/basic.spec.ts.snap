// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic functionality: a formatted table 1`] = `
"| **A** | **B** | **C** |
| ----: | :---- | ----- |
|     a | b     | c     |
|     x | y     | z     |
"
`;

exports[`Basic functionality: alternating keeps information of different codification 1`] = `
"* a
* b
  * c

  * d
    * e

    * f
"
`;

exports[`Basic functionality: dont break obsidian links 1`] = `
"[[respect this]]
"
`;

exports[`Basic functionality: enumerate lists 1`] = `
"1. foo

a. aaa
b. bbb
c. ccc

1. bar
2. baz
"
`;

exports[`Basic functionality: enumerate lists 2`] = `
"1. foo
   1. aaa
   2. bbb
   3. ccc
2. bar
3. baz
"
`;

exports[`Basic functionality: lists use * bullets 1`] = `
"* a
* b
  * c
  * d
    * e
    * f
"
`;

exports[`Basic functionality: lists use - bullets 1`] = `
"- a
- b
  - c
  - d
    - e
    - f
"
`;

exports[`Embed image 1`] = `
"<https://cyberculturesblog.files.wordpress.com/2019/01/Memex.jpg>
"
`;

exports[`Embed youtube 1`] = `
"<https://www.youtube.com/watch?v=B6rKUf9DWRI>
"
`;

exports[`Indent size 1`] = `
"- foo
- foo
- foo
- foo
"
`;

exports[`Indent size 2`] = `
"- foo
- foo
- foo
- foo
"
`;

exports[`Newlines around headings: disabled 1`] = `
"# Heading
Content
## Subheading
"
`;

exports[`Newlines around headings: enabled 1`] = `
"# Heading

Content

## Subheading
"
`;
